### PR TITLE
style(lint): switch to airbnb style guide

### PR DESCRIPTION
**Pull request by jackcmeyer**
**Changes proposed in this pull request:**
- switch to airbnb style guide
- ensured linter works with vscode autoformatting

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,20 +1,43 @@
...
+  globals: {
+    Atomics: 'readonly',
+    SharedArrayBuffer: 'readonly',
+  },
   parser: '@typescript-eslint/parser',
   parserOptions: {
-    project: './tsconfig.json',
-    tsconfigRootDir: './',
-    // TODO: we need this because of an issue with @typescript-eslint/parser: https://github.com/typescript-eslint/typescript-eslint/issues/864
-    createDefaultProgram: true,
```
Comment (fox1t) [inline .eslintrc.js]: Why you removed this lines? We need them if we want IDE support. It is a eslint bug and this is the only way to make it check newly created files without restarting VSCode.

File: .eslintrc.js [hunk 2]
```diff
@@ -1,20 +1,43 @@
...
+    'airbnb',
+    'prettier',
+  ],
+  globals: {
+    Atomics: 'readonly',
+    SharedArrayBuffer: 'readonly',
+  },
   parser: '@typescript-eslint/parser',
   parserOptions: {
-    project: './tsconfig.json',
```
Comment (fox1t) [inline .eslintrc.js]: We need this to let know eslint typescript parser how to use rules that depends on types: this is different from eslint for javascript where there are no rules that depends on types.

File: .eslintrc.js [hunk 3]
```diff
@@ -1,20 +1,43 @@
...
-    createDefaultProgram: true,
+    ecmaFeatures: {
+      jsx: true,
+    },
+    ecmaVersion: 2018,
+    sourceType: 'module'
   },
-  extends: [
-    'react-app',
-    'plugin:@typescript-eslint/recommended',
```
Comment (fox1t) [inline .eslintrc.js]: This is also a needed plugin in order to make eslint working properly with typescript.

File: package.json [hunk 4]
```diff
@@ -18,9 +18,10 @@
     "dev": "tsdx watch",
     "build": "tsdx build",
     "prepublishOnly": "npm run build",
-    "pretest": "npm run lint && npm run build",
     "test": "tsdx test --env=jsdom",
```
Comment (fox1t) [inline package.json]: If test is testing the compiled files we need to run build script before, otherwise we are going to test old files.

File: package.json [hunk 5]
```diff
@@ -18,9 +18,10 @@
...
-    "lint": "prettier --write --loglevel warn \"src/**/*\" \"stories/**/*\" \"test/**/*\" \"README.md\" \"docs/**/*\" && tsdx lint --fix \"src/**/*.{ts,tsx}\" \"test/**/*.{ts,tsx}\" \"stories/**/*.{ts,tsx}\"",
+    "prelint": "npm run prettier",
+    "lint": "tsdx lint \"src/**/*.{ts,tsx}\" \"test/**/*.{ts,tsx}\" \"stories/**/*.{ts,tsx}\"",
```
Comment (fox1t) [inline package.json]: Here is missing `--fix` option.

File: package.json [hunk 6]
```diff
@@ -49,16 +50,6 @@
       "git add"
     ]
   },
-  "prettier": {
```
Comment (fox1t) [inline package.json]: Unfortunately we need both configurations: this one is the only one that could be read by tsdx. If you want you can remove the one in .prettier file, but then we will have some problems with VSCode. The configuration that was there, was wisely crafted to work in all scenarios (and it failed anyway :( )

**Comment by fox1t** [comment]
@irvelervel and @tehKapa check this out and tell me if this new configuration is working for you

File: .prettierignore [hunk 7]
```diff
@@ -0,0 +1 @@
+./src/index.ts
```
Comment (fox1t) [inline .prettierignore]: Why we need this?

**Comment by fox1t** [comment]
There is a missing dev-dependency: `Error: Failed to load config "airbnb" to extend from.` that prevents CI passing.

**Comment by tehKapa** [comment]
> There is a missing dev-dependency: `Error: Failed to load config "airbnb" to extend from.` that prevents CI passing.

I've the same issue on local with `yarn lint` command

File: package.json (hunk 4 above)
Comment (jackcmeyer) [reply package.json]: I just made an obvious bug, and without rebuilding, reran the test suite and the tests failed. 

Are there edge cases where this won't always works?

File: package.json (hunk 5 above)
Comment (jackcmeyer) [reply package.json]: I was having issues with the `-fix` command producing code that didn't compile. 

Maybe this was due to the missing configuration above.. I will try again with that config.

File: .eslintrc.js (hunk 3 above)
Comment (jackcmeyer) [reply .eslintrc.js]: I will add it back 👍

File: .eslintrc.js (hunk 2 above)
Comment (jackcmeyer) [reply .eslintrc.js]: I will add it back 👍

File: package.json (hunk 4 above)
Comment (fox1t) [reply package.json]: Ok, then here we are testing ./src folder and not the compiled code. It is fine, however there could be cases in which ./src works and ./dist doesn't. For example types compilation problems: src works because TS used for tests infers and read correctly the types from the srource code but the compiled .d.ts files are somehow wrong. We can leave it as is, but we must be aware of this edge case. It will surely popup at some point.

File: package.json (hunk 5 above)
Comment (fox1t) [reply package.json]: Yea, it has to compile, otherwise our linting is broken.

Comment (jackcmeyer) [reply package.json]: `--fix has been added back`

File: .prettierignore (hunk 7 above)
Comment (jackcmeyer) [reply .prettierignore]: Removed. 

Prettier was formatting in a way that caused a bug in this file, however, this has been fixed was some configurations.

**Comment by jackcmeyer** [comment]
> > There is a missing dev-dependency: `Error: Failed to load config "airbnb" to extend from.` that prevents CI passing.
> 
> I've the same issue on local with `yarn lint` command

Fixed this issue.

File: .eslintrc.js (hunk 1 above)
Comment (jackcmeyer) [reply .eslintrc.js]: This has been added back.

File: .eslintrc.js (hunk 2 above)
Comment (jackcmeyer) [reply .eslintrc.js]: This has been added back.

File: package.json (hunk 6 above)
Comment (jackcmeyer) [reply package.json]: This has been added back in.

File: .prettierignore (hunk 7 above)
Comment (fox1t) [reply .prettierignore]: OK. Nice.

File: .eslintrc.js [hunk 8]
```diff
@@ -1,20 +1,53 @@
...
+    'react/jsx-one-expression-per-line': 'off',
+    'react/jsx-wrap-multilines': 'off',
+    'no-param-reassign': ['error', { 'props': false }],
+    'import/prefer-default-export': 'off',
+    '@typescript-eslint/member-delimiter-style': 'off',
     '@typescript-eslint/explicit-function-return-type': 'off',
     '@typescript-eslint/no-explicit-any': 'off',
   },
-}
+};
```
Comment (fox1t) [inline .eslintrc.js]: semi-comma?

Comment (jackcmeyer) [reply .eslintrc.js]: Removed

**Comment by jackcmeyer** [comment]
@StefanoMiC your changes look great! Thanks for the teamwork on this PR.

**Comment by tehKapa** [comment]
> @StefanoMiC your changes look great! Thanks for the teamwork on this PR.
> 
> I checked it out and verified that it works as expected.

I'm guessing you wanted to quote @irvelervel :D
